### ISSUE TITLE
Breadcrumb settings

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#2351]
One thing that's always frustrated me with Drupal is how themes are responsible for breadcrumbs. Some themes have settings, like whether to hide breadcrumbs if there's just a Home link, whether to show the current page title at the end of the trail, whether to show a separator at the end of the trail (for displaying the breadcrumbs just above/before the page title), etc. And then some themes don't have any settings and you have to do all that yourself in theme overrides.

What I'd like to see in Backdrop is these settings added to the breadcrumb block, separate from themes. Like the settings mentioned above, and also what character(s) to use for the separator (that's currently hard-coded in `theme_breadcrumb()`).

**Comment by olafgrabienski** [comment]
There are already some related issues:

- https://github.com/backdrop/backdrop-issues/issues/2260
- https://github.com/backdrop/backdrop-issues/issues/1535
- https://github.com/backdrop/backdrop-issues/issues/1509

But I guess this feature request is the most comprehensive. And in my opinion, such settings would be very helpful!

**Comment by vstemen** [comment]
Hi.

It was recommended that I post in this issue about this.

I have been working on a simple CSS only theme that relies fully on Backdrop's 
core layout system for layout.  I am developing it for our own business sites
but I am also preparing it to contribute to Backdrop.  This seems to me to be 
the idea behind backdrops architecture, to be able to write themes by styling
only, with no PHP template code needed.  I feel this is one of Backdrops great
advantages over Drupal.  Interestingly, I have not found any existing themes
that fully utilize this feature of Backdrop and are written only using CSS.

It has been very successful so far, except for one thing.  When I started
playing with styling the breadcrumb, after pulling my hair out trying to
figure out why I could not remove or replace the separator in CSS,
I discovered it is apparently hard coded in the PHP code in 
core/includes/theme.inc.  Since the breadcrumb separator is something that CSS
is designed to style, I feel this should corrected.  I assume it is a carry
over from Drupal that just has not been updated yet to fit with backdrop's
layout separation architecture.

I am attempting to revive this discussion in hopes that I can motivate the
powers that be to up the priority on this issue.

**Comment by klonos** [comment]
I think that we should:

1. remove the hard-coded setting and move it to css that can be overridden
2. make the breadcrumb character configurable via the breadcrumb block settings

**Comment by klonos** [comment]
...tentatively marking this as a bug (for the bit about not being able to override it easily) + task (for making the separator character a block setting)

**Comment by opi** [comment]
> 1. remove the hard-coded bit and move it to css that can be overridden
> 2. make the breadcrumb character configurable via the breadcrumb block settings

If the separator is CSS only (with `.breadcrumb li::before{}` I guess), then it can't be configurable via a settings. But it's not a problem. An accessible breadcrumb should be an ordered list (`<ol>`) of links wrapped inside a `<nav>` tag. Source https://www.w3.org/TR/wai-aria-practices/examples/breadcrumb/index.html
Anything else, including the separator, is presentation and should be only CSS.

One interesting setting is "[ ] Show current page's title", and this could be done by adding/removing the `element-invisible` class on the last item.

**Comment by klonos** [comment]
Thanks @BWPanda 👍 ...I gave this a quick spin in the PR sandbox, and everything seems to be working just fine.

There is one small issue with the breadcrumb when it comes to the home page and the respective layout, where even if the "Hide the breadcrumb trail if the only breadcrumb is 'Home'" setting is disabled, the only way to have the breadcrumb actually be rendered is to select "Current page title" from the "Append to the breadcrumb trail" option. But I think that this might be pre-existing, and not related to this PR specifically. I have no recommendations at the moment; just mentioning this.

I like the flexibility that we are adding by allowing the separator to be specified in the breadcrumb block, because as a new Drupal/Backdrop user that's where I'd expect to find such an option. Having said that, if one wishes to change the breadcrumb separator, they'd most likely do it site-wide (not saying that there will not be use cases that would need separate settings for separate sections/layouts - just not an 80% case). Having that setting live in the breadcrumb block though means that they'd have to change it for every single instance of that block in every layout they have placed it. This might not be much of an issue in small sites with a couple of layouts, but b.org for example has a couple of dozen layouts.

Again, I don't have any specific recommendation for the last issue; just mentioning it here. Possible solutions could be:

- Make the breadcrumb separator a *single* theme setting, but still have it "embedded" in the breadcrumb block. So every other setting in the breadcrumb block settings dialog would be saved in a respective `layout.layout.[layout_name].json` file, but this specific setting would be saved in `[theme_name].settings.json`. But the question is how do we know which theme? ...or do we save in the theme currently set as the default?

- Instead of having the separator setting be a field in the breadcrumb block settings dialog, we could have some help text that indicates what character is set for the current active theme, and have a link to the theme settings page., But what happens if the currently active theme does not have a settings page?

- Make the breadcrumb block a "reusable" block. Indicate that changing the settings in a single layout would update them across the rest of the layouts, except the default admin layout. Again, making an assumption here that there would only be a single layout for all admin pages.

I would like to know what others think.

**Comment by BWPanda** [comment]
Thanks for the review @klonos! Regarding the separator issue, as far as I know, you've just described the issue with Layouts/Blocks that we're trying to address with https://github.com/backdrop/backdrop-issues/issues/2245. The same problem is present in other blocks (e.g. the Header block has options to display the logo, site name, etc., and presumably you'd want these same setting across all layouts; the same with the Primary Navigation block, you'd likely want the same type of menu style across all layouts (top-level, drop-down or hierarchical)). So to me this is a core Layouts issue and not specific to breadcrumbs/the separator/this PR. Feel free to correct me if I'm wrong though :smiley:

**Comment by klonos** [comment]
Not wrong at all @BWPanda ...I did actually have that ticket in mind, but it seems to me that it will take some time before we actually have that implemented in core (given that that ticket has not seen any activity in 2 years now).

Good point 👍 ...and kudos for successfully using that as an argument to not block this issue 😄

**Comment by herbdool** [comment]
t works fine but it still might be better to leave this to CSS in the theme. We could just update a core CSS file to include it by default.

**Comment by vstemen** [comment]
I thought about it some more.  I don't know how difficult this would be to
implement in backdrop, but here is what I think would be the best solution
from a user perspective.

In addition to the new separator field in the breadcrumb block
configuration, add a check box, that is off by default, that says to
override the separator defined in the theme.  If it is set, output the
block CSS after the theme CSS.  If it is not set then output the theme CSS
last so that it will override the block setting.  If the theme does not
try to change the separator, then the checkbox will have no effect.

This way, themes that want to change it can by default, yet the user can
still override the theme with their own separator choice by checking that
box.

**Comment by jenlampton** [comment]
This issue still needs more feedback, removing milestone, but adding milestone candidate tag.

**Comment by vstemen** [comment]
If it would be quicker and easier for now, to solve the theming problem, my fallback solution would be to just keep the current block configuration dialog as it is.  Don't add a setting for the separator.  Just change it so that the current default separator is output via CSS rather than HTML so we can override it via our custom theme CSS or using CSS Injector.

My guess is that would be pretty easy to implement.  We could always add settings for it in the block configuration dialog later.

Would that be practical to keep it in the current milestone?

**Comment by jenlampton** [comment]
> Would that be practical to keep it in the current milestone?

1.11.4 was released about an hour ago 🎉 The soonest this could go in would be 1.12.1. 

> Don't add a setting for the separator. Just change it so that the current default separator is output via CSS

I love this idea!

**Comment by BWPanda** [comment]
I've updated the [PR](https://github.com/backdrop/backdrop/pull/2388/files) with suggestions from @opi and @herbdool, namely:

- Removing the separator from the block settings and moving it to CSS instead
- Always adding the current page title but hiding it with `element-invisible` as necessary

Please test and report your feedback :slightly_smiling_face:

**Comment by klonos** [comment]
I understand the benefit of the latest changes, but if a wear my PMC hat, as a constituent for non-coders/site-builders, I am ambivalent re this change. Referring to [this](https://backdropcms.org/philosophy#importance):

> **Backdrop values the needs of the editor and architect over the needs of the developer.**
>
> It's most important that the product first be usable, and second, easy to build and extend. Developers should consider the editorial experience first, the architect experience second, and put the needs of other developers last. Intuitive user interfaces are more valuable than extensive flexibility. When options are presented in the user interface, they should be made as clear and concise as possible.

According to the above, we should be favouring the setting in the block over the CSS implementation (which is code).

**Comment by herbdool** [comment]
On the other hand, it's not great UX to have a lot of options. Too many choices confuse and irritate people. I think WordPress' philosophy is decisions over options, or something like that. I think most people won't care about a breadcrumb separator, but perhaps that's my bias.

**Comment by BWPanda** [comment]
Ok, I had an idea: why not please everyone? (except @herbdool, sorry) I've re-worked the [PR](https://github.com/backdrop/backdrop/pull/2388) to add back in the UI settings for the separator, but then output it via CSS so it can be overridden easily by themes.

**Comment by olafgrabienski** [comment]
@BWPanda In the sandbox site, the breadcrumb block separator setting isn't respected at the moment. Maybe related: *Status report* says there are database updates (don't know if I should run them).

**Comment by BWPanda** [comment]
@olafgrabienski The can can override it, and most core themes do this by default. I tested with my own custom theme, not sure how to test in the sandbox... Maybe just check the CSS?

**Comment by olafgrabienski** [comment]
@BWPanda Re sandbox site testing: the site reflects the PR. So, if you go to the sandbox site (and change configuration to test), the PR *should* work as expected.

In your PR's sandbox site, Bartik is set as default theme. In the breadcrumb block configuration, I see `>>>ooo` in the *Separator* input, but the breadcrumb on the *About* page displays the default separator instead.

**Comment by BWPanda** [comment]
@olafgrabienski OK, so I can see what you're saying. I've tried both Bartók and Seven, but as I'm on my phone I can't check the CSS to see if those themes are setting their own separator. But I suspect that's the case and its designed that way.

Unless you're saying the UI setting should override the theme...?

**Comment by olafgrabienski** [comment]
@BWPanda Thanks for your feedback! Maybe I was a bit naive, or I didn't read well all comments of the issue.

When most or many (core) themes do override the breadcrumb separator and when consequently the breadcrumb separator configuration option usually wont't have any effect (unless I add some CSS), I find it quite confusing to see the option in the UI. I guess, in this case I don't agree to the current approach and would prefer one of the other approaches (UI setting with effect, or CSS only).

**Comment by BWPanda** [comment]
Hmm, I guess this brings up the question: should UI settings override the theme, or should the theme override UI settings?

Wonder if that's been discussed previously...

**Comment by vstemen** [comment]
Hi.  I previously discussed it.

If you have not already seen it, please see my comment about this
https://github.com/backdrop/backdrop-issues/issues/2351#issuecomment-445984784

My other comment for a quicker more immediate solution sounds like what BWPanda did
https://github.com/backdrop/backdrop-issues/issues/2351#issuecomment-454548910

I haven't had a chance to test it yet myself.

**Comment by BWPanda** [comment]
Yes, reading back through this issue it seems that some (including @vstemen and @klonos) have suggested that the UI should override the theme.

In regards to @vstemen's suggestion though, what if we make it so that if the UI separator is blank, the theme controls it, and if it's set, it overrides the theme? Would that work for everyone?

**Comment by vstemen** [comment]
That would probably work OK, but I still think I like my original suggestion of 
"add a checkbox, that is off by default, that says to override the separator
defined in the theme.", for several reasons.

1.  And this is the main reason, with the checkbox off, you can have the
default separator configured that is always used unless the theme overrides it, 
If you have to leave the separator blank, as you suggest, if you have a theme
that does not define it's own separator, it will mess up the breadcrumb unless
you know to go back in and define a separator.

2.  It allows having a default separator that stays configured in the UI that
can be enabled or disabled for different themes.  This is useful, especially for
special characters like that little '>>' character that has no keyboard key to
enter it, you can just re-enable it with the checkbox.

3.  The enable/disable theme override checkbox in the UI makes it very clear to
the user whether the theme has control of it or not.

**Comment by BWPanda** [comment]
@vstemen All good points :smiley: 
Let's see what others think...

**Comment by jenlampton** [comment]
> should UI settings override the theme, or should the theme override UI settings?

I think we should be very carful not to create one instance where things work differently than they do everywhere else. Everywhere else, the theme always gets the last say. If we created one place where the breadcrumb CSS were to come after the theme CSS, that would frustrate all the theme developers (and many site architects too).

The important thing, though, is not to create a UI that is broken, or one that depends on having special CSS in a theme.

Here is the current `theme_breadcrumb()` function:

```php
function theme_breadcrumb($variables) {
  $breadcrumb = $variables['breadcrumb'];
  $output = '';
  if (!empty($breadcrumb)) {
    $output .= '<nav role="navigation" class="breadcrumb">';
    // Provide a navigational heading to give context for breadcrumb links to
    // screen-reader users. Make the heading invisible with .element-invisible.
    $output .= '<h2 class="element-invisible">' . t('You are here') . '</h2>';
    $output .= '<ol><li>' . implode(' » </li><li>', $breadcrumb) . '</li></ol>';
    $output .= '</nav>';
  }
  return $output;
}
```
In order for this change not to break existing sites, I don't think we can switch from a markup separator to a CSS separator. There might be something fancy we can do by adding different classes to the `OL` tag, but IMHO the risk of our CSS not working with existing contrib/custom CSS is too high when we start changing the markup underneath. 

If we still wanted to make the breadcrumb separator configurable (which, for the record, I'm not sure is an 80% use-case -- I'd personally prefer that this be a contrib module) we'd need to make the ` » ` part into a variable.

Another thought is that maybe this shouldn't be a block setting, since we probably don't want a different breadcrumb separator on each page.

**Comment by vstemen** [comment]
> I think we should be very carful not to create one instance where things work
> differently than they do everywhere else. Everywhere else, the theme always
> gets the last say. If we created one place where the breadcrumb CSS were to
> come after the theme CSS, that would frustrate all the theme developers (and
> many site architects too).

Yes.  I agree.  Which is why I recommended having the checkbox that says to
override the separator defined in the theme "**off**" by default in the UI block.
Only if you enable that would it be able to override a themes CSS separator
setting.

It seems to me that would give us the best of both worlds without breaking
expected behavior by default.

Although, I could live with or without that feature in the breadcrumb block UI.
Without it, we should at least have Backdrop's default separator output via CSS
before any theme or CSS Injector CSS so that it can be overridden.  In my
opinion that would at east solve the current problem of Backdrop hard coding
the separator so that my theme cannot change it.

**Comment by jenlampton** [comment]
> Which is why I recommended having the checkbox that says to
override the separator defined in the theme "off" by default in the UI block.

The problem is anything at all in the UI that would override the theme. That happens nowhere else, and should not be introduced here.

> It seems to me that would give us the best of both worlds without breaking
expected behavior by default.

I'd prefer to not break expected behavior at all :)

> Without it, we should at least have Backdrop's default separator output via CSS
before any theme or CSS Injector CSS so that it can be overridden.

Or not use CSS at all. Like we do currently.

> In my opinion that would at east solve the current problem of Backdrop hard coding
the separator so that my theme cannot change it.

What? Your theme **can** change it. Override `theme_breadcrumb()` in template.php. If you don't want the character in the markup, remove it from your theme's copy of `theme_breadcrumb()` and add it via CSS instead.

**Comment by vstemen** [comment]
> Your theme can change it: Override theme_breadcrumb() in template.php. If you
> don't want the character in the markup, remove it from your theme's copy of
> theme_breadcrumb() and add it via CSS instead.

I think you are looking at it from more of a programmers point of view.  The
idea was to have the ability to have a CSS only theme, without having to modify
template code, like I mentioned early in this thread back in November 2018
https://github.com/backdrop/backdrop-issues/issues/2351#issuecomment-439128208

To me, one of the beautys of Backdrop's layout system is the ability to create
CSS only themes with no PHP coding as I did with the Corporate KISS theme
I contributed.

Also, I was expecting to be able to change the breadcrumb separator in CSS
Injector.  As I mentioned early on, I pulled my hair out initially trying to
figure out why none of my CSS was able change the separator.  I am not clear why
it would need to be hard coded in the template PHP code.

If you don't like the idea of adding any controls for it in the breadcrumb
block, I would be happy with the simple CSS solution I mentioned in 
https://github.com/backdrop/backdrop-issues/issues/2351#issuecomment-454548910.

You responded back then that you loved the idea.  I am in hopes that you will
still feel that way :-).

**Comment by BWPanda** [comment]
For posterity...

> I guess this brings up the question: should UI settings override the theme, or should the theme override UI settings?

In https://github.com/backdrop/backdrop-issues/issues/3822 we realised/decided that "**themes can always override UI settings**" (https://github.com/backdrop/backdrop-issues/issues/3822#issuecomment-505842098).

**Comment by BWPanda** [comment]
PR updated. Let's try and get this reviewed/tested...

**Comment by BWPanda** [comment]
@olafgrabienski It's a bit hard by default, since the core themes seem to override the breadcrumb settings ([as is their right](https://github.com/backdrop/backdrop-issues/issues/3822#issuecomment-505842098)), but you can play around with the breadcrumb block, see the new settings form, and see what options you can get working. You can also try installing a different theme that maybe doesn't override any breadcrumb stuff...

**Comment by olafgrabienski** [comment]
Thanks @BWPanda. I've played around with the breadcrumb settings in the PR sandbox. In *Basis*, the settings work nicely, with the exception of the separator (characters to display between breadcrumb items).

Overview re Basis:

- `Separator` - doesn't work
- `Add the current page title to the end of the breadcrumb trail` - works √
- `Hide the breadcrumb trail if the only breadcrumb is 'Home'` - works √

As *Basis* is the default theme, and one important setting doesn't apply, shall we create a follow-up regarding the Separator setting in Basis?

**Comment by indigoxela** [comment]
A word of warning: I suspect these settings won't work at all with any theme (contrib or custom) that overrides theme_breadcrumb().

**Comment by herbdool** [comment]
I created [related issue](https://github.com/backdrop/backdrop-issues/issues/6034) and PR to address the accessibility of the separator. This issue is still relevant if there's consensus, though wrapping the separator in a span with a class will also allow it to be hidden with CSS, which was one thing brought up here.
